### ISSUE TITLE
Missing fields with 0.16 and unions

### ISSUE THREAD

**Issue by elado** [rmosolgo/graphql-ruby#190]
Here's an example repo: https://github.com/elado/graphql-ruby-0.16-bug

Relay queries use fragments, `fragment X` and `...X`. In 0.16 there seems to be a bug with unions where the first type doesn't get some fields in the result. This didn't happen on 0.15.

Here are the queries and the results:

Flat query works:

![Flat query works](https://cloud.githubusercontent.com/assets/51488/17004515/2bea6eba-4e8a-11e6-8dfe-4dc078ce201c.png)

But a broken query is missing the `name` field of `EntityAaa`:

![Fragment query is missing fields](https://cloud.githubusercontent.com/assets/51488/17004514/2bea7252-4e8a-11e6-8113-13eb006d8324.png)

The flat query:

```graphql
query {
  viewer {
    collections {
      edges {
        node {
          id
          name
          items {
            edges {
              node {
                id
                entity {
                  __typename
                  ... on EntityAaa {
                    id
                    name
                    aaa
                    entityType
                    __typename
                  }
                  ... on EntityBbb {
                    id
                    name
                    bbb
                    entityType
                    __typename
                  }
                }
              }
            }
          }
        }
      }
    }
  }
}
```

Fragment query:

```graphql
query {
  viewer {
    collections {
      edges {
        node {
          id
          name
          items {
            edges {
              node {
                id
                entity {
                  __typename
                  ...F1
                  ...F2
                }
              }
            }
          }
        }
      }
    }
  }
}

fragment F1 on EntityAaa {
  id
  name
  aaa
  entityType
  __typename
}

fragment F2 on EntityBbb {
  id
  name
  bbb
  entityType
  __typename
}
```

Changing gem to `gem 'graphql', '< 0.16'` will cause both queries to work.

**Comment by elado** [comment]
I'm assuming it's related to #180 ?

**Comment by rmosolgo** [comment]
Yes I'm sure you're right, let me dig in!

**Comment by rmosolgo** [comment]
Just pushed 0.16.1 with a fix for this (56e55b1)! Thanks for the great test case, that made it really easy to find. 

I was correctly merging type info when the fields were discovered inline, but just as you pointed out, type info was lost when merging fragments. 

Could you try 0.16.1? 

One more todo: I want to make sure interface types also work in this case, but I have to run out for a bit!

**Comment by rmosolgo** [comment]
I suspect there are some little bugs in places where `irep_node.definition` is used, since it only captures the _first_ field definition that it finds. For example, it could apply wrong default arguments in places where the same field occurs twice, or it could use the wrong query complexity value. 

I'll fix it in the morning!

**Comment by elado** [comment]
Seems like the missing fields are back! Thanks!

**Comment by rmosolgo** [comment]
Thanks again for the report, just shipped 0.17 which includes some fixes for interfaces + query complexity in a similar case!
